rust/nasl-builtin-regex/src/lib.rs

```diff
@@ -0,0 +1,196 @@
+// SPDX-FileCopyrightText: 2024 Greenbone AG
+//
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+use nasl_builtin_utils::{error::FunctionErrorKind, NaslFunction};
+use nasl_builtin_utils::{Context, Register};
+use nasl_function_proc_macro::nasl_function;
+use nasl_syntax::NaslValue;
+use regex::{Regex, RegexBuilder};
+
+fn parse_search_string<'a>(s: &'a mut &str, rnul: bool, multiline: bool) -> &'a str {
+    if !rnul {
+        *s = s.split('\0').next().unwrap();
+    }
+    if !multiline {
+        *s = s.split('\n').next().unwrap();
+    }
+
+    s
+}
+
+fn make_regex(pattern: &str, icase: bool, multiline: bool) -> Result<Regex, FunctionErrorKind> {
+    match RegexBuilder::new(pattern.to_string().as_str())
+        .case_insensitive(icase)
+        .multi_line(multiline)
+        .build()
+    {
+        Ok(re) => Ok(re),
+        Err(e) => Err(FunctionErrorKind::Dirty(format!(
+            " Error building regular expression pattern: {}",
+            e
+        ))),
+    }
+}
+
+/// Matches a string against a regular expression.
+/// - string  String to search the pattern in
+/// - pattern the pattern that should be matched
+/// - icase   case insensitive flag
+/// - rnul    replace the null char in the string. Default TRUE.
+/// - multiline Is FALSE by default (string is truncated at the first
+///                “end of line”), and can be set to TRUE for multiline search.
+/// Return true if matches, false otherwise
+#[nasl_function(named(string, pattern, icase, rnul, multiline))]
+fn ereg(
+    string: NaslValue,
+    pattern: NaslValue,
+    icase: Option<bool>,
+    rnul: Option<bool>,
+    multiline: Option<bool>,
+) -> Result<bool, FunctionErrorKind> {
+    let icase = icase.unwrap_or(false);
+    let rnul = rnul.unwrap_or(true);
+    let multiline = multiline.unwrap_or(false);
+
+    let str_bind = string.to_string();
+    let mut str_aux = str_bind.as_str();
+
+    let str_aux = parse_search_string(&mut str_aux, rnul, multiline);
+
+    let re = make_regex(&pattern.to_string(), icase, multiline)?;
+    Ok(re.is_match(str_aux))
```

Comment:
With the change to `parse_search_string` above, we can simplify this even further and get rid of `str_aux` and `str_bind`:
```rust
    let string = string.to_string();
    let string = parse_search_string(&string, rnul, multiline);

    let re = make_regex(&pattern.to_string(), icase, multiline)?;
    Ok(re.is_match(string))
```
Doing it this way avoids the explicit call to `.as_str()` (which is needed only rarely). 

Small digression: The reason this works is that the compiler will see that we are calling a function (`parse_search_string`) with a `&String` argument, but the function takes a `&str`. What happens next is that the compiler checks the `Deref` impl of `String`. `String` implements the `Deref` trait with `Target = &str`, and so the compiler knows that it can turn a `&String` into a `&str`, which it will automatically do. This is a pretty confusing part of Rust but quite important for lots of convenient syntax. This magic is called deref coercion and happens automatically in function calls - see https://doc.rust-lang.org/std/ops/trait.Deref.html#deref-coercion